File: backend/routers/comparator.py
"""F02 · Comparateur de prompts"""
from fastapi import APIRouter
from pydantic import BaseModel
from services.diff_engine import compute_diff

router = APIRouter()


class DiffRequest(BaseModel):
    left: str
    right: str
    mode: str = "auto"


@router.post("/diff")
def diff_prompts(req: DiffRequest):
    return compute_diff(req.left, req.right, req.mode)


@router.post("/similarity")
def similarity(req: DiffRequest):
    result = compute_diff(req.left, req.right, req.mode)
    return {
        "similarity_jaccard": result["similarity_jaccard"],
        "similarity_levenshtein": result["similarity_levenshtein"],
        "common_count": len(result["common"]),
        "left_only_count": len(result["left_only"]),
        "right_only_count": len(result["right_only"]),
    }

File: frontend/src/api/comparator.ts
import { api } from './client'

export interface DiffOp {
  op: 'equal' | 'insert' | 'delete'
  tokens: string[]
}

export interface DiffResult {
  similarity_jaccard: number
  similarity_levenshtein: number
  left_only: string[]
  right_only: string[]
  common: string[]
  left_token_count: number
  right_token_count: number
  diff: DiffOp[]
}

export const comparatorApi = {
  diff: (left: string, right: string, mode = 'auto') =>
    api.post<DiffResult>('/comparator/diff', { left, right, mode }).then((r) => r.data),

  similarity: (left: string, right: string, mode = 'auto') =>
    api.post('/comparator/similarity', { left, right, mode }).then((r) => r.data),
}
